pkg/types/validation/installconfig.go

```diff
@@ -99,6 +101,115 @@ func ValidateInstallConfig(c *types.InstallConfig, openStackValidValuesFetcher o
 	if _, ok := validPublishingStrategies[c.Publish]; !ok {
 		allErrs = append(allErrs, field.NotSupported(field.NewPath("publish"), c.Publish, validPublishingStrategyValues))
 	}
+
+	return allErrs
+}
+
+// IPAddressTypeByField is a map of field path to whether they request IPv4 or IPv6.
+type IPAddressTypeByField map[string]struct{ IPv4, IPv6 bool }
+
+// IPByField is a map of field path to the net.IPs in sorted order.
+type IPByField map[string][]net.IP
+
+// InferIPVersionFromInstallConfig infers the user's desired ip version from the networking config.
+// Presence field names match the field path of the struct within the Networking type. This function
+// assumes a valid install config.
+func InferIPVersionFromInstallConfig(n *types.Networking) (hasIPv4, hasIPv6 bool, presence IPAddressTypeByField, addresses IPByField) {
+	if n == nil {
+		return
+	}
+	addresses = make(IPByField)
+	for _, network := range n.MachineNetwork {
+		addresses["machineNetwork"] = append(addresses["machineNetwork"], network.CIDR.IP)
+	}
+	for _, network := range n.ServiceNetwork {
+		addresses["serviceNetwork"] = append(addresses["serviceNetwork"], network.IP)
+	}
+	for _, network := range n.ClusterNetwork {
+		addresses["clusterNetwork"] = append(addresses["clusterNetwork"], network.CIDR.IP)
+	}
+	presence = make(IPAddressTypeByField)
+	for k, ips := range addresses {
+		sort.Slice(ips, func(i, j int) bool { return bytes.Compare(ips[i], ips[j]) < 0 })
```

Comment:
golang sometimes represents IPv4 addresses as 4 bytes, and sometimes as 16 bytes using the IPv6-compatible representation of IPv4 addresses. While you *probably* only have addresses in the 16-byte format here, it might be better to guarantee that (eg by calling `.To16()` on each one as you fill in `addresses` above).

Reply:
is there any case where it won't be one or the other?

Reply:
(where it would be both)?

Reply:
`net.ParseIP` always returns 16-byte addresses, but other code between there and here might have decided to replace some of the 16-byte addresses with their `.To4()` forms. I don't have any specific reason for thinking that might have happened here, it's just that the direct byte comparison looks wrong to me since it's not safe in the fully-general case.

Reply:
In the net module Go is using To4 to check IPv4 ness

```
// DefaultMask returns the default IP mask for the IP address ip.
// Only IPv4 addresses have default masks; DefaultMask returns
// nil if ip is not a valid IPv4 address.
func (ip IP) DefaultMask() IPMask {
	if ip = ip.To4(); ip == nil {
		return nil
	}
```

If someone gives us the IPv6 representation of an IPv4, (::hhhh:hhhh), they'd have a zero CIDR prefix.  We could probably just check for that, but that sounds... rare.

Reply:
Sorry, what I'm trying to say is that `net.IP` has two different-but-equivalent representations of IPv4 addresses, and if you're looking at the `[]byte` representation directly rather than using `net.IP` APIs, you might not get the answers you expect:

    ip := net.ParseIP("1.2.3.4")

    ip.Equal(ip.To4())        → true
    bytes.Equal(ip, ip.To4()) → false

Reply:
him, according to this code and code we have in Kube To4() means "a 4 address" - so if those other places are wrong we need to fix them too.